tests/functional/test_alias.py

```diff
@@ -0,0 +1,147 @@
+
+    def test_subcommand_alias_with_global_params(self):
+        self.add_alias(
+            'my-alias',
+            'ec2 describe-regions --query Regions[].RegionName --output text')
+        self.parsed_responses = [
+            {
+                'Regions': [
+                    {
+                        'Endpoint': 'ec2.us-east-1.amazonaws.com',
+                        'RegionName': 'us-east-1'
+                    }
+                ]
+            }
+        ]
+        cmdline = 'my-alias'
+        stdout, _, _ = self.assert_params_for_cmd(cmdline, {})
+        self.assertEqual(len(self.operations_called), 1)
+        self.assertEqual(self.operations_called[0][0].name, 'DescribeRegions')
+        self.assertEqual(stdout.strip('\n'), 'us-east-1')
+
+    def test_subcommand_alias_then_global_params(self):
+        self.add_alias('my-alias', 'ec2 describe-regions')
+        self.parsed_responses = [
+            {
+                'Regions': [
+                    {
+                        'Endpoint': 'ec2.us-east-1.amazonaws.com',
+                        'RegionName': 'us-east-1'
+                    }
+                ]
+            }
+        ]
+        cmdline = 'my-alias '
+        cmdline += '--query=Regions[].RegionName '
+        cmdline += '--output=text'
+        stdout, _, _ = self.assert_params_for_cmd(cmdline, {})
+        self.assertEqual(len(self.operations_called), 1)
+        self.assertEqual(self.operations_called[0][0].name, 'DescribeRegions')
+        self.assertEqual(stdout.strip('\n'), 'us-east-1')
+
+    def test_global_params_then_subcommand_alias(self):
+        self.add_alias('my-alias', 'ec2 describe-regions')
+        self.parsed_responses = [
+            {
+                'Regions': [
+                    {
+                        'Endpoint': 'ec2.us-east-1.amazonaws.com',
+                        'RegionName': 'us-east-1'
+                    }
+                ]
+            }
+        ]
+        cmdline = '--query=Regions[].RegionName '
+        cmdline += '--output=text '
+        cmdline += 'my-alias'
+        stdout, _, _ = self.assert_params_for_cmd(cmdline, {})
+        self.assertEqual(len(self.operations_called), 1)
+        self.assertEqual(self.operations_called[0][0].name, 'DescribeRegions')
+        self.assertEqual(stdout.strip('\n'), 'us-east-1')
+
+    def test_alias_overrides_builtin_command(self):
+        self.add_alias('ec2', 's3api')
+        cmdline = 'ec2 list-buckets'
+        self.assert_params_for_cmd(cmdline, {})
+        self.assertEqual(len(self.operations_called), 1)
+        self.assertEqual(self.operations_called[0][0].name, 'ListBuckets')
+
+    def test_external_alias(self):
+        # The external alias is tested by using mkdir; a command that
+        # is universal for the various OS's we support
+        directory_to_make = os.path.join(self.files.rootdir, 'newdir')
+        self.add_alias('mkdir', '!mkdir %s' % directory_to_make)
```

Comment:
So I tested this on Windows to make sure that command will work. But does anyone have strong feelings about using a subprocess in a functional test. I feel it is sort of a gray area, but if anyone feels strongly about it, I can move it to an integration test. I made it a functional test because those get run more than integration tests and otherwise I am not too sure how I could test this functionally without patching out the subprocess.call, which would leak implementation detail.